### PR TITLE
refactor: Introduce `Fragmentizer`

### LINKED ISSUE #129: Literal parsing has O(n²) complexity
If `ReceiveState` parses a message with n literals, then the complexity is O(n²). There is already a TODO in `receive.rs`:

```rust
// Try to parse the whole message from the start (including the new line).
// TODO: If the message is really long and we need multiple attempts to receive it, then this is O(n^2)
//       IMO this can be only fixed by using a generator-like decoder
match self.codec.decode(&self.read_buffer[..self.seen_bytes]) {
    Ok((remaining, message)) => {
        assert!(remaining.is_empty());
        Ok(Some(ReceiveEvent::DecodingSuccess(message.into_static())))
    }
    Err(error) => Ok(Some(ReceiveEvent::DecodingFailure(error.into_static()))),
}
```

Unfortunately I think there is no easy fix for that. The parser implemented in `imap-codec` would need to be stateful and remember the partially parsed message in order to solve the issue.

### LINKED ISSUE #198: feat: Support secure discard of invalid messages
Related https://github.com/duesee/imap-codec/issues/432#issuecomment-1962427538

### PR DESCRIPTION
**Pull request by jakoschiko**
In the long term we might want to add a utility to `imap-codec` for better message/line/literal parsing. But before that, let's do some experiments in this repository.

The module `receive` already contains some primitive line/literal parsing. Let's extract and extend this functionality.

Closes #129
Closes #198

### CODE REVIEW

File: Cargo.toml [hunk 1]
```diff
@@ -15,11 +15,11 @@ exclude = [
 [features]
 default = ["stream"]
 expose_stream = []
-stream = ["dep:rustls", "dep:tokio", "dep:tokio-rustls"]
+stream = ["dep:bytes", "dep:rustls", "dep:tokio", "dep:tokio-rustls"]
```
[the remark this answers is not in the document]
Comment (jakoschiko) [reply Cargo.toml:18]: `Stream` [is still using it](https://github.com/duesee/imap-next/blob/c0dd3e1c513bb5a7c789e98cfce11a5b7ca5523c/src/stream.rs#L24-L25) for interacting with `TcpStream`.

[the remark this answers is not in the document]
Comment (duesee) [reply Cargo.toml:18]: Ah, my bad.

File: src/lib.rs [hunk 2]
```diff
@@ -1,12 +1,12 @@
 #![forbid(unsafe_code)]
 
 pub mod client;
-mod client_receive;
 mod client_send;
+#[doc(hidden)] // TODO: move to imap-codec and make it public
```
Comment (duesee) [inline src/lib.rs:5]: We could create an issue and reference it here.

File: src/fragmentizer.rs [hunk 3]
```diff
@@ -0,0 +1,1411 @@
...
+#[derive(Clone, Debug)]
+enum Parser {
+    Line(LineParser),
+    Literal(LiteralParser),
+}
+
+/// Stateful parser for the next line fragment.
+#[derive(Clone, Debug)]
+struct LineParser {
+    // Where we started parsing the line.
```
Comment (duesee) [inline src/fragmentizer.rs:325]:
```suggestion
    /// Where we started parsing the line.
```

File: src/fragmentizer.rs [hunk 4]
```diff
@@ -0,0 +1,1411 @@
...
+    start: usize,
+    // Until where we parsed the line.
```
Comment (duesee) [inline src/fragmentizer.rs:327]:
```suggestion
    /// Until where we parsed the line.
```

File: src/fragmentizer.rs [hunk 5]
```diff
@@ -0,0 +1,1411 @@
...
+    end: usize,
+    // Accumulated state based on the parsed bytes.
```
Comment (duesee) [inline src/fragmentizer.rs:329]:
```suggestion
    /// Accumulated state based on the parsed bytes.
```

File: src/fragmentizer.rs [hunk 6]
```diff
@@ -0,0 +1,1411 @@
...
+    },
+    Cr {
+        announcement: Option<LiteralAnnouncement>,
+    },
+}
+
+/// Stateful parser for the next literal fragment.
+#[derive(Clone, Debug)]
+struct LiteralParser {
+    // Where we started parsing the literal.
```
Comment (duesee) [inline src/fragmentizer.rs:497]:
```suggestion
    /// Where we started parsing the literal.
```

File: src/fragmentizer.rs [hunk 7]
```diff
@@ -0,0 +1,1411 @@
...
+    start: usize,
+    // Until where we parsed the literal.
```
Comment (duesee) [inline src/fragmentizer.rs:499]:
```suggestion
    /// Until where we parsed the literal.
```

File: src/fragmentizer.rs [hunk 8]
```diff
@@ -0,0 +1,1411 @@
...
+    end: usize,
+    // Remaining bytes we need to parse.
```
Comment (duesee) [inline src/fragmentizer.rs:501]:
```suggestion
    /// Remaining bytes we need to parse.
```

**Review by duesee — APPROVED** [review]
Thanks! I am very happy with the changes! Spend a bit of time reviewing the tests, and also implemented some fuzz testing (on another branch). I also added a server example that fits here well, and tested if it works as it should.

I propose to merge this and to make a second PR to add fuzz testing and improve the tests. Fuzz testing can easily be moved from imap-next to imap-codec. So, we don't need to worry too much when to bring it in place.

Great work!

File: src/lib.rs (hunk 2 above)
Comment (jakoschiko) [reply src/lib.rs:5]: I've created #238

**Pull request merged**